library/ssl_srv.c

```diff
@@ -204,28 +219,48 @@ static int ssl_parse_signature_algorithms_ext( mbedtls_ssl_context *ssl,
         return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_HELLO );
     }
 
-    /*
-     * For now, ignore the SignatureAlgorithm part and rely on offered
-     * ciphersuites only for that part. To be fixed later.
+    /* Currently we only guarantee signing the ServerKeyExchange message according
+     * to the constraints specified in this extension (see above), so it suffices 
+     * to remember only one suitable hash for each possible signature algorithm. 
      *
-     * So, just look at the HashAlgorithm part.
+     * This will change when we also consider certificate signatures,
+     * in which case we will need to remember the whole signature-hash
+     * pair list from the extension. 
      */
-    for( md_cur = ssl->conf->sig_hashes; *md_cur != MBEDTLS_MD_NONE; md_cur++ ) {
-        for( p = buf + 2; p < end; p += 2 ) {
-            if( *md_cur == (int) mbedtls_ssl_md_alg_from_hash( p[0] ) ) {
-                ssl->handshake->sig_alg = p[0];
-                goto have_sig_alg;
-            }
-        }
-    }
 
-    /* Some key echanges do not need signatures at all */
-    MBEDTLS_SSL_DEBUG_MSG( 3, ( "no signature_algorithm in common" ) );
-    return( 0 );
+    for( p = buf + 2; p < end; p += 2 ) {
+
+        /* Silently ignore unknown signatures or such where
+         * we have already found a suitable hash algorithm. */
```

Comment:
Thanks, the second part of the comment is outdated. We ignore unknown signatures, but we _are_ in principle interested in the full list of supported hash algorithms per signature. The current implementation keeps only one because there's not yet the need for more, but there will be once the sig-hash extension is also taken into account in the choice of server certificate.